lib/Sema/CSSimplify.cpp

```diff
@@ -4036,8 +4036,19 @@ ConstraintSystem::simplifyKeyPathConstraint(Type keyPathTy,
       // get any special power from being formed in certain contexts, such
       // as the ability to assign to `let`s in initialization contexts, so
       // we pass null for the DC to `isSettable` here.)
-      if (!storage->isSettable(nullptr)
-          || !storage->isSetterAccessibleFrom(DC)) {
+      if (!getASTContext().isSwiftVersionAtLeast(5)) {
+        // As a source-compatibility measure, continue to allow
+        // WritableKeyPaths to be formed in the same conditions we did
+        // in previous releases even if we should not be able to set
+        // the value in this context.
+        if (!storage->isSettable(DC)) {
```

Comment:
This should be `nullptr`, not `DC`. (Stored `let` properties are settable within the constructor but not in general.)

Reply:
I'm intentionally restoring the code that we previously shipped for the sake of source compatibility.

The Swift 5 check below this does pass `nullptr` here.

Reply:
I believe this one'll crash at runtime, though, unlike checking for access control.

Reply:
I guess an API could have taken a WritableKeyPath but happened to only use it for reading, though…

Reply:
I don't know what's changed since [SR-6667](https://bugs.swift.org/browse/SR-6667) was opened, but DI seems to be complaining about writes through the keypath now:

```
error: address of constant 'self.bar' taken before it is initialized
    self[keyPath: path] = true
```

If you initialize the property before this point, you get:

```
error: immutable value 'self.bar' must not be passed inout
    self[keyPath: path] = true
```

So it seems that anyone who actually did try to write through it will now see this error rather than a crash.

Reply:
I see. But that also means you might as well have it do the correct thing, right?

Reply:
Well in theory someone could be caching the key path and only using it to read, as you mentioned, or even caching it and using it to write, e.g.:

```
struct S {
  let bar: Bool
  let path: WritableKeyPath<S, Bool>

  init() {
    path = \S.bar
    bar = true
  }

  mutating func clear() {
    let t = path
    self[keyPath: t] = false
  }
}
```